Dockerfile

```diff
@@ -1,80 +1,41 @@
 
 RUN cd /tinygo/ && \
-    go install /tinygo/
+    make llvm-source
 
-# tinygo-wasm stage installs the needed dependencies to compile TinyGo programs for WASM.
-FROM tinygo-base AS tinygo-wasm
-
-COPY --from=tinygo-base /go/bin/tinygo /go/bin/tinygo
-COPY --from=tinygo-base /tinygo/src /tinygo/src
-COPY --from=tinygo-base /tinygo/targets /tinygo/targets
+# tinygo-llvm-build stage build the custom llvm with xtensa support
+FROM tinygo-llvm AS tinygo-llvm-build
 
 RUN cd /tinygo/ && \
-    apt-get update && \
-    apt-get install -y make clang-11 libllvm11 lld-11 cmake ninja-build && \
-    mkdir build && \
-    make wasi-libc binaryen
+    make llvm-build
 
-# tinygo-avr stage installs the needed dependencies to compile TinyGo programs for AVR microcontrollers.
-FROM tinygo-base AS tinygo-avr
+# tinygo-compiler stage builds the compiler itself
+FROM tinygo-llvm-build AS tinygo-compiler
 
-COPY --from=tinygo-base /go/bin/tinygo /go/bin/tinygo
-COPY --from=tinygo-base /tinygo/src /tinygo/src
-COPY --from=tinygo-base /tinygo/targets /tinygo/targets
-COPY --from=tinygo-base /tinygo/Makefile /tinygo/
-COPY --from=tinygo-base /tinygo/tools /tinygo/tools
-COPY --from=tinygo-base /tinygo/lib /tinygo/lib
+COPY . /tinygo
 
+# remove submodules directories and re-init them to fix any hard-coded paths
+# after copying the tinygo directory in the previous step.
 RUN cd /tinygo/ && \
-    apt-get update && \
-    apt-get install -y apt-utils make binutils-avr gcc-avr avr-libc && \
-    make gen-device-avr && \
-    apt-get autoremove -y && \
-    apt-get clean
-
-# tinygo-arm stage installs the needed dependencies to compile TinyGo programs for ARM microcontrollers.
-FROM tinygo-base AS tinygo-arm
-
-COPY --from=tinygo-base /go/bin/tinygo /go/bin/tinygo
-COPY --from=tinygo-base /tinygo/src /tinygo/src
-COPY --from=tinygo-base /tinygo/targets /tinygo/targets
-COPY --from=tinygo-base /tinygo/Makefile /tinygo/
-COPY --from=tinygo-base /tinygo/tools /tinygo/tools
-COPY --from=tinygo-base /tinygo/lib /tinygo/lib
+    rm -rf ./lib/* && \
+    git submodule sync && \
+    git submodule update --init --recursive --force
 
 RUN cd /tinygo/ && \
-    apt-get update && \
-    apt-get install -y apt-utils make clang-11 && \
-    make gen-device-nrf && make gen-device-stm32
+    make
 
-# tinygo-all stage installs the needed dependencies to compile TinyGo programs for all platforms.
-FROM tinygo-wasm AS tinygo-all
+# tinygo-tools stage installs the needed dependencies to compile TinyGo programs for all platforms.
+FROM tinygo-compiler AS tinygo-tools
 
-COPY --from=tinygo-base /tinygo/Makefile /tinygo/
-COPY --from=tinygo-base /tinygo/tools /tinygo/tools
-COPY --from=tinygo-base /tinygo/lib /tinygo/lib
+COPY --from=tinygo-compiler /tinygo/build/tinygo /go/bin/tinygo
 
 RUN cd /tinygo/ && \
-    apt-get update && \
-    apt-get install -y apt-utils make clang-11 binutils-avr gcc-avr avr-libc && \
-    make gen-device
+    make wasi-libc binaryen gen-device
```

Comment:
It may be better to use something like "make gen-device -j".
The only difference is the speed.

Reply:
Is `-j` actually used by `gen-device` task? I did not see that flag has any effect on that, and `wasi-libc` already has a `-j4` in the makefile itself. Or am I missing something here?

Reply:
Yes, `gen-device` is sped up with the `-j` flag. It runs `gen-device-avr`, `gen-device-sam`, etc in parallel.

Reply:
OK, I can update that.

Reply:
Updated.